backend/x11/backend.c

```diff
@@ -116,14 +116,70 @@ static void handle_x11_event(struct wlr_x11_backend *x11,
 		}
 		break;
 	}
+	case XCB_UNMAP_NOTIFY: {
+		xcb_unmap_notify_event_t *ev = (xcb_unmap_notify_event_t *)event;
+		struct wlr_x11_output *output =
+			get_x11_output_from_window_id(x11, ev->window);
+		if (output != NULL) {
+			wlr_log(WLR_DEBUG, "Window unmapped, stopping updates");
+			output->mapped = false;
+		}
+		break;
+	}
+	case XCB_MAP_NOTIFY: {
+		xcb_unmap_notify_event_t *ev = (xcb_unmap_notify_event_t *)event;
+		struct wlr_x11_output *output =
+			get_x11_output_from_window_id(x11, ev->window);
+		if (output != NULL) {
+			wlr_log(WLR_DEBUG, "Window mapped%s", !output->hidden ? ", resuming updates" : "");
+			output->mapped = true;
+			if (!output->hidden) {
+				wlr_output_send_frame(&output->wlr_output);
+			}
+		}
+		break;
+	}
+	case XCB_PROPERTY_NOTIFY: {
+		xcb_property_notify_event_t *ev = (xcb_property_notify_event_t *)event;
+		struct wlr_x11_output *output =
+			get_x11_output_from_window_id(x11, ev->window);
+		if (output == NULL) {
+			break;
+		}
+		if (ev->atom == x11->atoms.net_wm_state) {
+			xcb_get_property_cookie_t cookie =
+				xcb_get_property(x11->xcb, false, ev->window,
+					ev->atom, XCB_ATOM_ATOM, 0, 32);
+			xcb_get_property_reply_t *reply =
```

Comment:
The spec says:

> The Window Manager should remove the property whenever a window is withdrawn

So we probably need to handle the case where the property is removed, and reset `hidden` to false. `xcb_property_notify_event_t.state` indicates whether a property is deleted.